hawkbit-dmf/hawkbit-dmf-amqp/src/main/java/org/eclipse/hawkbit/amqp/AmqpFormulateErrorMessage.java

```diff
@@ -0,0 +1,44 @@
+/**
+ * Copyright (c) 2021 Bosch.IO GmbH and others.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.eclipse.hawkbit.amqp;
+
+import org.springframework.amqp.core.Message;
+import org.springframework.amqp.rabbit.support.ListenerExecutionFailedException;
+
+import java.util.Collection;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * Class that formulates a meaningful error message
+ */
+public class AmqpFormulateErrorMessage {
+
+    /**
+     * Constructs an error message based on failed message content
+     *
+     * @param throwable
+     *                  the throwable containing failed message content
+     * @return
+     *                  meaningful error message
+     */
+    public String constructErrorMessage(final Throwable throwable) {
+        StringBuilder completeErrorMessage = new StringBuilder();
+
+        //extract failed message content
+        Collection<Message> failedMessages = ((ListenerExecutionFailedException) throwable).getFailedMessages();
+        Message failedMessage = failedMessages.iterator().next();
+        Map<String, Object> messageProperties = failedMessage.getMessageProperties().getHeaders();
+        String properties = messageProperties.keySet().stream().map(key -> key + "=" + messageProperties.get(key))
+                .collect(Collectors.joining(", ", "{", "}"));
+
+        completeErrorMessage.append(throwable.getCause().getMessage()).append(new String(failedMessage.getBody())).append(properties);
```

Comment:
I would decomposition a method into logical steps, so at the end we would have something like `completeErrorMessage.append(mainErrorMsg).append(amqpMsgBody).append(amqpMsgHeaders).toString()`. I would be much more readable this way :)